include/bluetooth/mesh/access.h

```diff
@@ -257,14 +257,12 @@ struct bt_mesh_model_op {
 #define BT_MESH_MODEL_CB(_id, _op, _pub, _user_data, _cb)                    \
 {                                                                            \
 	.id = (_id),                                                         \
-	.op = _op,                                                           \
-	.keys = { [0 ... (CONFIG_BT_MESH_MODEL_KEY_COUNT - 1)] =             \
-			BT_MESH_KEY_UNUSED },                                \
 	.pub = _pub,                                                         \
-	.groups = { [0 ... (CONFIG_BT_MESH_MODEL_GROUP_COUNT - 1)] =         \
-			BT_MESH_ADDR_UNASSIGNED },                           \
-	.user_data = _user_data,                                             \
+	.keys = { BT_MESH_KEY_UNUSED },                                      \
```

Comment:
This isn't doing the same thing as the original code. `BT_MESH_KEY_UNUSED` is defined as `0xffff`, however what you're doing here is only initialising `.keys[0]` to that value, whereas all other values of the array get initialised to zero.

Reply:
Thanks; I've never used that syntax and didn't realize it was a spread operation.  If so, there may be no way to make this header C++ compatible.

Reply:
It's a GNU C extension: https://gcc.gnu.org/onlinedocs/gcc/Designated-Inits.html

However, it's a fair question to ask whether we should allow such extensions or require standard C code everywhere. It's super convenient though and probably saves a bit of CPU time since the array doesn't need to be initialised at runtime.

Reply:
I haven't tested it out, but I think the designated initializer can be replaced by some of the macro utils (include/sys/util.h) that we have.

Something like:
```
#define FOO(i, _) BT_MESH_ADDR_UNASSIGNED ,
.groups = { UTIL_LISTIFY(CONFIG_BT_MESH_MODEL_GROUP_COUNT - 1, FOO) }
```
I'm not expert on these though, so you might find a better way in there.

Reply:
@joerchan Thanks, that seems to work.  It does require defining a helper macro to generate the constant; maintainers please provide feedback on how that should be spelled and hidden from the user.

One approach not done here is to move the helper functions to the bottom of the file, outside the doxygen block.  This works, but may be confusing.

Reply:
I think @trond-snekvik is the best person to answer the naming and hiding question. I suppose this should probably be re-usable by users for their own models and so it shouldn't be hidden.

Reply:
OK.  I provided a new one that puts all the internals above the documentation block.  The one with "public" (no `_` affix) could be moved into the block and documented, if it provides something of use outside the helper macros.